libs/wingc/src/lifting.rs

```diff
@@ -0,0 +1,360 @@
+					),
+				});
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	fn should_capture_expr(&self, node: &Expr) -> bool {
+		let ExprKind::Reference(ref r) = node.kind else {
+		return false;
+	};
+
+		match r {
+			Reference::Identifier(ref symb) => {
+				let fullname = symb.name.clone();
+				let span = symb.span.clone();
+				self.should_capture_reference(&fullname, &span)
+			}
+			Reference::TypeReference(ref t) => {
+				let fullname = t.full_path_str();
+				let span = t.span.clone();
+
+				// skip "This" (which is the type of "this")
+				if let Some(class) = &self.ctx.current_class() {
+					if class.full_path_str() == fullname {
+						return false;
+					}
+				}
+
+				self.should_capture_reference(&fullname, &span)
+			}
+			_ => false,
+		}
+	}
+
+	fn jsify_expr(&self, node: &Expr, phase: Phase) -> String {
+		self.jsify.jsify_expression(
+			&node,
+			&mut JSifyContext {
+				in_json: false,
+				phase: phase,
+				files: &mut Files::default(),
+				lifts: &mut Lifts::disabled(),
+			},
+		)
+	}
+}
+
+impl<'a> Fold for LiftTransform<'a> {
+	fn fold_reference(&mut self, node: Reference) -> Reference {
+		match node {
+			Reference::InstanceMember {
+				object,
+				property,
+				optional_accessor,
+			} => {
+				self.ctx.push_property(property.name.clone());
+				let result = Reference::InstanceMember {
+					object: Box::new(self.fold_expr(*object)),
+					property: self.fold_symbol(property),
+					optional_accessor,
+				};
+				self.ctx.pop_property();
+				return result;
+			}
+			Reference::TypeMember { typeobject, property } => {
+				self.ctx.push_property(property.name.clone());
+				let result = Reference::TypeMember {
+					typeobject: Box::new(self.fold_expr(*typeobject)),
+					property: self.fold_symbol(property),
+				};
+				self.ctx.pop_property();
+				return result;
+			}
+			_ => {}
+		}
+
+		fold::fold_reference(self, node)
+	}
+
+	fn fold_expr(&mut self, node: Expr) -> Expr {
+		let expr_phase = self.jsify.types.get_expr_phase(&node).unwrap();
+		let expr_type = self.jsify.types.get_expr_type(&node).unwrap();
+		let is_field = includes_this(&node);
+
+		// this whole thing only applies to inflight expressions
+		if self.ctx.current_phase() == Phase::Preflight {
+			return fold::fold_expr(self, node);
+		}
+
+		// if this expression represents the current class, no need to capture it (it is by definition
+		// available in the current scope)
+		if self.is_self_type_reference(&node) {
+			return fold::fold_expr(self, node);
+		}
+
+		//---------------
+		// LIFT
+
+		if expr_phase == Phase::Preflight {
+			// jsify the expression so we can get the preflight code
+			let code = self.jsify_expr(&node, Phase::Preflight);
+
+			// check that we can qualify the lift (e.g. determine which property is being accessed)
+			if self.ctx.current_property().is_none()
+				&& expr_type.as_preflight_class().is_some()
+				&& !expr_type.is_handler_preflight_class()
```

Comment:
I'm wondering regarding `!expr_type.is_handler_preflight_class()`. A handler preflight class might also be a normal preflight class. In which case it can be used in such a way that'll also make it not possible to qualify (no property access). The exception might be when calling this class. So perhaps we  need to make sure this is a "call" expression and only then make it an exception?